lib/leanplum_api/api.rb

```diff
@@ -222,9 +223,17 @@ def extract_user_id_or_device_id_hash!(hash)
     # As of 2015-10 Leanplum supports ISO8601 date & time strings as user attributes.
     def build_user_attributes_hash(user_hash, action = 'setUserAttributes')
       user_hash = HashWithIndifferentAccess.new(user_hash)
```

Comment:
you need to allow for devices to be at the top level of the hash. This is fixed in my 4.x fork. For reference.

Reply:
Also this method is doing too many things. Consider breaking this method up into much smaller chunks.

Reply:
lol you've seen the actual `setUserAttributes` signature?  it's like the definition of "doing too many things".  but i feel you...  not sure it's worth trying to address here though given that it's addressed on your fork.